endpoints/openrtb2/auction.go

```diff
@@ -653,17 +654,40 @@ func mapSChains(req *openrtb_ext.RequestWrapper) error {
 	return nil
 }
 
-func validateAndFillSourceTID(req *openrtb2.BidRequest) error {
+func validateAndFillSourceTID(req *openrtb_ext.RequestWrapper, generateRequestID bool, hasStoredBidRequest bool, isAmp bool) error {
 	if req.Source == nil {
 		req.Source = &openrtb2.Source{}
 	}
-	if req.Source.TID == "" {
-		if rawUUID, err := uuid.NewV4(); err == nil {
-			req.Source.TID = rawUUID.String()
-		} else {
-			return errors.New("req.Source.TID missing in the req and error creating a random UID")
+
+	rawUUID, err := uuid.NewV4()
+	if err != nil {
+		return errors.New("error creating a random UUID for source.tid")
+	}
+
+	if req.Source.TID == "" || req.Source.TID == "{{UUID}}" {
+		req.Source.TID = rawUUID.String()
+	}
+
+	if generateRequestID && (isAmp || hasStoredBidRequest) {
+		req.Source.TID = rawUUID.String()
+	}
+
+	for _, impWrapper := range req.GetImp() {
+		ie, _ := impWrapper.GetImpExt()
+		rawUUID, err := uuid.NewV4()
+		if err != nil {
+			return errors.New("imp.ext.tid missing in the imp and error creating a random UID")
+		}
+		if ie.GetTid() == "" || ie.GetTid() == "{{UUID}}" {
```

Comment:
Shouldn't we also check that `req.Source.Tid` was originally empty before setting `imp.ext.tid` as per the spec in this PR? Probably need to set a boolean to indicate if this was empty before the code above is executed, or move this above setting `req.Source.Tid`.

Reply:
Sorry my PR summary was outdated, now updated. The logic comes from Bret's comment on this issue if you want to cross reference. We went away from tying source.tid to imp.ext.tid 

https://github.com/prebid/prebid-server/issues/2381#issuecomment-1284533924

Reply:
Okay, if we are adhering to what prebid.js is already doing, then I am okay. There is a bit of tension between allowing publishers to send in IDs while also leveraging stored requests for config management vs. dealing with IDs that were inadvertently put into a stored request rendering them static.